More quantum circuit library refactoring

alexanderivrii opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

As a part of circuit library refactoring #13046, more quantum circuits now have their "modern" variants,
either as gates or as synthesis methods.

This PR adds: 
- `GraphState` -> `GraphStateGate`
- `FourierChecking` -> `fourier_checking(...)`
- `UnitaryOverlap` -> `unitary_overlap(...)`
- `HiddenLinearFunction` -> `hidden_linear_function(...)`
- `PhaseEstimation` -> `phase_estimation(...)`

### Details and comments

For now these are all in Python.

alexanderivrii committed:
GraphState -> GraphStateGate

alexanderivrii committed:
adding GraphStateGate::__eq__ method

alexanderivrii committed:
oops... fourier checking should not be a gate

alexanderivrii committed:
FourierChecking -> fourier_checking

alexanderivrii committed:
UnitaryOverlap -> unitary_overlap

alexanderivrii committed:
visualization improvements (following quantum_volume code)

alexanderivrii committed:
visualization improvements (following quantum_volume code)

alexanderivrii committed:
HiddenLinearFunction -> hidden_linear_function

alexanderivrii committed:
unused imports

alexanderivrii committed:
PhaseEstimation -> phase_estimation

alexanderivrii committed:
cleaning up phase estimation code

alexanderivrii committed:
reno

Cryoris commented on qiskit/circuit/library/fourier_checking.py:
```diff
@@ -72,26 +78,62 @@ def __init__(self, f: List[int], g: List[int]) -> None:
+
+    circuit = QuantumCircuit(int(num_qubits), name=f"fc: {f}, {g}")
+    circuit.h(circuit.qubits)
+    circuit.compose(Diagonal(f), inplace=True)
```
Can we directly use `DiagonalGate` here? `Diagonal` is a circuit that just uses `DiagonalGate` internally 🙂

alexanderivrii replied:
Great suggestion! Fixed. However, because this changes how the circuit looks like, I have reverted to use the old code for `FourierChecking` circuit.

Cryoris commented on qiskit/circuit/library/graph_state.py:
```diff
@@ -84,3 +90,87 @@ def __init__(self, adjacency_matrix: list | np.ndarray) -> None:
+    **Reference Circuit:**
+
+    .. plot::
+
```
This applies to all other docstrings as well: What do you think about including the source to show how this object can be used? (I think you need to re-indent the code with this flag, so better not accept this suggestion)
```suggestion
    .. plot::
        :include-source:
```

alexanderivrii replied:
Done here and in all other files.

Cryoris commented on qiskit/circuit/library/graph_state.py:
```diff
@@ -10,13 +10,14 @@
+from qiskit.utils.deprecation import deprecate_func
 
 
 class GraphState(QuantumCircuit):
```
Can we make this circuit use the gate internally?

alexanderivrii replied:
Great suggestion! Done.

Cryoris commented on qiskit/circuit/library/hidden_linear_function.py:
```diff
@@ -77,22 +83,78 @@ def __init__(self, adjacency_matrix: Union[List[List[int]], np.ndarray]) -> None
+        circuit = hidden_linear_function(adjacency_matrix)
         super().__init__(*circuit.qregs, name=circuit.name)
-        self.compose(circuit.to_gate(), qubits=self.qubits, inplace=True)
+        self.compose(circuit, qubits=self.qubits, inplace=True)
```
We probably still want this wrapping to keep `HiddenLinearFunction` unchanged
```suggestion
        self.append(circuit.to_gate(), self.qubits)
```

alexanderivrii replied:
Fixed.

Cryoris commented on qiskit/circuit/library/overlap.py:
```diff
@@ -41,13 +42,13 @@ class UnitaryOverlap(QuantumCircuit):
         from qiskit.circuit.library import EfficientSU2, UnitaryOverlap
         from qiskit.primitives import Sampler
 
-        # get two circuit to prepare states of which we comput the overlap
```
Nice catches!

Cryoris commented on qiskit/circuit/library/phase_estimation.py:
```diff
@@ -97,3 +102,70 @@ def __init__(
+
+    circuit.append(QFTGate(num_evaluation_qubits).inverse(), qr_eval[:])
+
+    reversal_pattern = list(reversed(range(num_evaluation_qubits)))
```
Could you verify that transpiling the circuit will correctly remove all Swaps? 🙂

alexanderivrii replied:
Yes, I have added the test.

Cryoris reviewed (commented):
Looks great, thanks for taking care of these! I left some minor comments below.

alexanderivrii committed:
Merge branch 'main' into finalize_modernize

alexanderivrii committed:
Restoring to the original definition of the FourierChecking circuit

alexanderivrii committed:
Merge branch 'finalize_modernize' of github.com:alexanderivrii/qiskit-terra into finalize_modernize

alexanderivrii committed:
pass over fourier_checking function

alexanderivrii committed:
remaining suggestions from code review; missing tests; missing API refs

alexanderivrii committed:
pylint

alexanderivrii commented:
Thanks @Cryoris, I have addressed the suggestions from the code review, and also added some missing tests, and added the functions to the API as in the other PRs.

alexanderivrii committed:
reno fix

Cryoris commented on test/python/circuit/library/test_phase_estimation.py:
```diff
@@ -149,6 +150,94 @@ def test_phase_estimation_iqft_setting(self):
+
+    def test_phase_estimation_function_swaps_removed(self):
+        """Test that transpiling the circuit correctly removes swaps and permutations."""
+        with self.subTest("U=S, psi=|1>"):
```
This subtest is probably an artifact 🙂

alexanderivrii replied:
oops, indeed, fixed in [36e3450](https://github.com/Qiskit/qiskit/pull/13353/commits/36e3450bf2f14eccffd5fcb574e9fb78f2c06d98)

Cryoris commented on qiskit/circuit/library/fourier_checking.py:
```diff
@@ -81,17 +87,72 @@ def __init__(self, f: List[int], g: List[int]) -> None:
         self.compose(circuit.to_gate(), qubits=self.qubits, inplace=True)
+
+
+def fourier_checking(f: List[int], g: List[int]) -> QuantumCircuit:
```
It would be nice to use `collections.abc.Sequence` in consistency with `DiagonalGate`, but we can also do this later 🙂 
```suggestion
def fourier_checking(f: Sequence[int], g: Sequence[int]) -> QuantumCircuit:
```

alexanderivrii replied:
sure, done in [36e3450](https://github.com/Qiskit/qiskit/pull/13353/commits/36e3450bf2f14eccffd5fcb574e9fb78f2c06d98)

alexanderivrii committed:
another small round of addressing review comments

Cryoris reviewed (approved):
Looks great, thank you! 🙂
